drivers: usb: udc: add Smartbond UDC driver

kasjer opened the pull request:
Code adds USB UDC driver for Smartbond device.

Code was NOT tested with ISO endpoints yet.
It was tested with console, mass, and cdc_acm applications

Additionally small fix in DMA power management was needed.

Function dma_smartbond_set_channel_status() used incorrect
condition to release state lock.
In initialization function dma_smartbond_init() function
dma_smartbond_set_channel_status() was called for each DMA
channel and tried to release lock that was never taken.

Fixes #74662 

Signed-off-by: Jerzy Kasenberg <jerzy.kasenberg@codecoup.pl>

kasjer committed:
drivers: usb: udc: add Smartbond UDC driver

Code adds Smartbond UDC driver to be used with
USB next stack.

Signed-off-by: Jerzy Kasenberg <jerzy.kasenberg@codecoup.pl>

kasjer committed:
drivers: dma: smartbond: Fix power policy handling

Function dma_smartbond_set_channel_status() used incorrect
condition to release state lock.
In initialization function dma_smartbond_init() function
dma_smartbond_set_channel_status() was called for each DMA
channel and tried to release lock that was never taken.

Signed-off-by: Jerzy Kasenberg <jerzy.kasenberg@codecoup.pl>

tmon-nordic commented on drivers/usb/udc/udc_smartbond.c:
```diff
@@ -0,0 +1,1784 @@
+
+	dev_state->dev = dev;
+
+	/*
+	 * You do not need to initialize it if your driver does not use
+	 * udc_lock_internal() / udc_unlock_internal(), but implements its
+	 * own mechanism.
+	 */
```
Looks like stale template comment that should be removed.

kasjer replied:
comment removed

ydamigos commented:
Please add  `usbd` as supported peripheral in DA1469x boards' yaml file.

kasjer committed:
boards: renesas: Smartbond: Add test supported feature usbd

Marks boards da1469x_dk_pro and da14695_dk_usb as supporting
new usb driver support

Signed-off-by: Jerzy Kasenberg <jerzy.kasenberg@codecoup.pl>

kasjer commented:
> Please add `usbd` as supported peripheral in DA1469x boards' yaml file.

done

ydamigos commented on drivers/usb/udc/udc_smartbond.c:
```diff
@@ -0,0 +1,1768 @@
+		USB->USB_FWMSK_REG &= ~(BIT(ep_idx - 1 + USB_USB_FWMSK_REG_USB_M_RXWARN31_Pos));
+	}
+
+	regs->rxc |= USB_USB_RXC1_REG_USB_RX_EN_Msk;
```
Could we reduce rxc register accesses?

kasjer replied:
donw

kasjer commented on drivers/usb/udc/udc_smartbond.c:
```diff
@@ -0,0 +1,1768 @@
+		fill_tx_fifo(ep_state);
+	}
+
+	regs->txc |= USB_USB_TXC1_REG_USB_TX_EN_Msk;
```
done

kasjer commented:
Build fails due to unrelated change #73189 (issue #79594)

jfischer-no commented on drivers/usb/udc/udc_smartbond.c:
```diff
@@ -0,0 +1,1764 @@
+	return 0;
+}
+
+/*
+ * This is called in the context of udc_ep_enqueue() and must
+ * not block. The driver can immediately claim the buffer if the queue is empty,
+ * but usually it is offloaded to a thread or work queue to handle transfers
+ * in a single location. Please refer to existing driver implementations
+ * for examples.
+ */
```
Please remove the comments copied from the udc_skeleton.c.

jfischer-no commented on boards/renesas/da14695_dk_usb/da14695_dk_usb.yaml:
```diff
@@ -9,9 +9,18 @@ toolchain:
   - xtools
 supported:
   - arduino_gpio
+  - counter
   - gpio
   - hwinfo
   - watchdog
   - i2c
   - spi
   - usb_device
+  - rtc
+  - crypto
+  - dma
+  - mipi_dbi
+  - display
+  - memc
```
It does not belong in this PR and you need to remove them from this commit. (These are test features)

jfischer-no commented on drivers/usb/udc/udc_smartbond.c:
```diff
@@ -0,0 +1,1764 @@
+
+LOG_MODULE_REGISTER(udc_smartbond, CONFIG_UDC_DRIVER_LOG_LEVEL);
+
+#define DT_DRV_COMPAT renesas_smartbond_usbd
+
+#define UDC_IRQ(inst)      DT_INST_IRQ_BY_IDX(inst, 0, irq)
+#define UDC_IRQ_PRI(inst)  DT_INST_IRQ_BY_IDX(inst, 0, priority)
+#define VBUS_IRQ(inst)     DT_INST_IRQ_BY_IDX(inst, 1, irq)
+#define VBUS_IRQ_PRI(inst) DT_INST_IRQ_BY_IDX(inst, 1, priority)
+
+/*
+ * Minimal transfer size needed to use DMA. For short transfers
+ * it may be simpler to just fill hardware FIFO with data instead
+ * of programming DMA registers.
+ */
+#define DMA_MIN_TRANSFER_SIZE(inst) DT_INST_PROP(inst, dma_min_transfer_size)
+#define FIFO_READ_THRESHOLD(inst)   DT_INST_PROP(inst, fifo_read_threshold)
```
Please move them to the bottom, just above `#define UDC_SMARTBOND_DEVICE_DEFINE(n) `

kasjer replied:
Suggestion applied

jfischer-no commented on drivers/usb/udc/udc_smartbond.c:
```diff
@@ -0,0 +1,1764 @@
+#define EP0_FIFO_SIZE 8
+#define EP_FIFO_SIZE  64
+
+#define EP_MAX DT_INST_PROP(0, num_bidir_endpoints)
+_Static_assert(EP_MAX > 0 && EP_MAX <= 4, "num-bidir-endpoints out of range");
```
Please move it to the bottom, above #define UDC_SMARTBOND_DEVICE_DEFINE(n). Also it does not take instance number in the account, which is probably okay on your platform, but still not correct. Use BUILD_ASSERT. The check is likely not necessary. There other ways to get number of endpoints.

jfischer-no commented on drivers/usb/udc/udc_smartbond.c:
```diff
@@ -0,0 +1,1764 @@
+	struct dma_config rx_cfg;
+	struct dma_block_config tx_block_cfg;
+	struct dma_block_config rx_block_cfg;
+} usbd_dma_cfg = {
+	.tx_chan = DT_DMAS_CELL_BY_NAME(DT_NODELABEL(usbd), tx, channel),
+	.tx_slot_mux = DT_DMAS_CELL_BY_NAME(DT_NODELABEL(usbd), tx, config),
+	.tx_dev = DEVICE_DT_GET(DT_DMAS_CTLR_BY_NAME(DT_NODELABEL(usbd), tx)),
+	.rx_chan = DT_DMAS_CELL_BY_NAME(DT_NODELABEL(usbd), rx, channel),
+	.rx_slot_mux = DT_DMAS_CELL_BY_NAME(DT_NODELABEL(usbd), rx, config),
+	.rx_dev = DEVICE_DT_GET(DT_DMAS_CTLR_BY_NAME(DT_NODELABEL(usbd), rx)),
+};
```
Please move the instantiation  of usbd_dma_cfg to the bottom, either as part of  UDC_SMARTBOND_DEVICE_DEFINE(n) or above it.

kasjer replied:
DMA configuration split into config (from DT) and data (that can change) and generated with along instance of device

jfischer-no commented on drivers/usb/udc/udc_smartbond.c:
```diff
@@ -0,0 +1,1764 @@
+
+static void udc_smartbond_isr(const struct device *dev)
+{
+	struct usb_dc_state *dev_state = dev->data;
```
Please name it `struct udc_smartbond_data;` and use `struct udc_smartbond_data *data = udc_get_private(dev);` or struct udc_smartbond_data *priv = udc_get_private(dev);` consistently.

kasjer replied:
dev_state change to data in all places

jfischer-no commented on drivers/usb/udc/udc_smartbond.c:
```diff
@@ -0,0 +1,1764 @@
+/**
+ * USB functionality can be disabled from HOST and DEVICE side.
+ * Host side is indicated by VBUS line.
+ * Device side is decided by pair of calls usb_dc_attach()/usb_dc_detach,
+ * USB will only work when application calls usb_dc_attach() and VBUS is present.
```
usb_dc_attach()?

kasjer replied:
change to reflect API calls udc_enable/udc_disable

kasjer commented on drivers/usb/udc/udc_smartbond.c:
```diff
@@ -0,0 +1,1764 @@
+			dev_state->setup_buffer[1], dev_state->setup_buffer[2],
+			dev_state->setup_buffer[3], dev_state->setup_buffer[4],
+			dev_state->setup_buffer[5], dev_state->setup_buffer[6],
+			dev_state->setup_buffer[7]);
```
Applied, thanks for suggestion

kasjer commented on drivers/usb/udc/udc_smartbond.c:
```diff
@@ -0,0 +1,1764 @@
+
+static void start_rx_packet(struct usb_dc_state *dev_state, struct smartbond_ep_state *ep_state)
+{
+	struct udc_ep_config *ep_config = &ep_state->config;
```
Suggestion applied

kasjer commented on drivers/usb/udc/udc_smartbond.c:
```diff
@@ -0,0 +1,1764 @@
+static void start_rx_packet(struct usb_dc_state *dev_state, struct smartbond_ep_state *ep_state)
+{
+	struct udc_ep_config *ep_config = &ep_state->config;
+	const struct udc_smartbond_config *cfg = dev_state->dev->config;
```
done

jfischer-no commented on drivers/usb/udc/udc_smartbond.c:
```diff
@@ -0,0 +1,1749 @@
+		remaining--;
+	}
+
+	if (ep_idx != 0) {
```
Wondering why you do not need to handle it for control endpoints?
Also, for constructs like this, you can just reverse the logic and just return. The code below will look a bit more relaxed.

```c
	if (ep_idx == 0) {
		/* Not needed for control endpoints because they are fancy. */
		return;
	}
```

kasjer replied:
Suggestion applied. Added explanation why this case does not apply. Additional check also added for cases where enabling FIFO level interrupt may not be needed.

jfischer-no commented on drivers/usb/udc/udc_smartbond.c:
```diff
@@ -0,0 +1,1749 @@
+static void start_rx_dma(const struct usb_smartbond_dma_config *dma_cfg, volatile void *src,
+			 void *dst, uint16_t size)
+{
+	if (dma_reload(dma_cfg->rx_dev, dma_cfg->rx_chan, (uint32_t)src, (uint32_t)dst, size) < 0) {
```
This line looks confusing. I doubt volatile above makes any sense. `uintptr_t` would be a more appropriate type.

kasjer replied:
Change as suggested. It was done this way because function start_rx_dma was called with address of volatile register.
Now casting is done in placed where start_rx_dma is called instead.

jfischer-no commented on drivers/usb/udc/udc_smartbond.c:
```diff
@@ -0,0 +1,1749 @@
+		.fifo_read_threshold = FIFO_READ_THRESHOLD(n),                                     \
+		.fifo_read_threshold = FIFO_READ_THRESHOLD(n),                                     \
+		.num_of_eps = DT_INST_PROP(n, num_bidir_endpoints),                                \
+		.dma_cfg =                                                                         \
+			{                                                                          \
+				.tx_chan = DT_INST_DMAS_CELL_BY_NAME(n, tx, channel),              \
+				.tx_slot_mux = DT_INST_DMAS_CELL_BY_NAME(n, tx, config),           \
+				.tx_dev = DEVICE_DT_GET(DT_INST_DMAS_CTLR_BY_NAME(n, tx)),         \
+				.rx_chan = DT_INST_DMAS_CELL_BY_NAME(n, rx, channel),              \
+				.rx_slot_mux = DT_INST_DMAS_CELL_BY_NAME(n, rx, config),           \
+				.rx_dev = DEVICE_DT_GET(DT_INST_DMAS_CTLR_BY_NAME(n, rx)),         \
+			},                                                                         \
```
```c
		.dma_cfg = {                                                                          \
			.tx_chan = DT_INST_DMAS_CELL_BY_NAME(n, tx, channel),              \
			.tx_slot_mux = DT_INST_DMAS_CELL_BY_NAME(n, tx, config),           \
			.tx_dev = DEVICE_DT_GET(DT_INST_DMAS_CTLR_BY_NAME(n, tx)),         \
			.rx_chan = DT_INST_DMAS_CELL_BY_NAME(n, rx, channel),              \
			.rx_slot_mux = DT_INST_DMAS_CELL_BY_NAME(n, rx, config),           \
			.rx_dev = DEVICE_DT_GET(DT_INST_DMAS_CTLR_BY_NAME(n, rx)),         \
		}, 
```

kasjer replied:
This strange formatting was done by `clang-format` tool. I wander if you have power to make someone change `.clang-format` to make it more consistent for nested structure initialization.

Fixed anyway.

jfischer-no commented on drivers/usb/udc/udc_smartbond.c:
```diff
@@ -0,0 +1,1749 @@
+	};                                                                                         \
+                                                                                                   \
+	static struct usb_smartbond_data udc_data_##n = {                                          \
+		.udc_data =                                                                        \
+			{                                                                          \
+				.mutex = Z_MUTEX_INITIALIZER(udc_data_##n.udc_data.mutex),         \
```
```c

		.udc_data = {                                                                          \
			.mutex = Z_MUTEX_INITIALIZER(udc_data_##n.udc_data.mutex), 
```

kasjer replied:
done

jfischer-no commented on drivers/usb/udc/udc_smartbond.c:
```diff
@@ -0,0 +1,1754 @@
+	LOG_INF("ep 0x%02x dequeue all", ep);
+
+	lock_key = irq_lock();
+
```
That looks too easy, do you need to find and cancel DMA transfers in progress?

kasjer replied:
Your are right, can you give me a quick tip on how to trigger this case so I can verify that it works correctly?

kasjer replied:
Fixed (scenario when this happens not tested though).

jfischer-no replied:
There are few, (There was a simple one using shell commands to enqueue transfers, I have to bring it back). You can try CDC ACM sample, build it with CONFIG_SHELL and CONFIG_USBD_SHELL enabled. On Zephyr side enter shell command `usbd select sample_usbd`.
```

python3
>>> import usb
>>> dev = usb.core.find(idVendor=0x2fe3, idProduct=0x0001)
>>> dev.detach_kernel_driver(0)
>>> #to set DTR
>>> dev.ctrl_transfer(0x21, 0x22, 3, 0, 0)
0
>>> #Read all DCD and DSR notifications
>>> dev.read(0x81, 100)
array('B', [161, 32, 0, 0, 0, 0, 2, 0, 1, 0])
>>> dev.read(0x81, 100)
array('B', [161, 32, 0, 0, 0, 0, 2, 0, 3, 0])
>>> #On Zephyr side 'usbd disable' should trigger dequeue transfers for endpoint 0x01
```
